data service: support OpenAPI Specification (#14291)
pingcap/docs: tidb-cloud/data-service-oas-with-nextjs.md

---
title: Use the OpenAPI Specification of a Data App with Next.js
summary: Learn how to use the OpenAPI Specification of a Data App to generate client code and develop a Next.js application.
---

# Use the OpenAPI Specification of a Data App with Next.js

## Before you begin

Before using OpenAPI Specification with Next.js, make sure that you have the following:

- A [TiDB Serverless cluster](/tidb-cloud/create-tidb-cluster-serverless.md)

> <!--TODO: add yarn-->

Reviewer: - Install node, [npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm) and [yarn](https://yarnpkg.com/getting-started/install)

## 1. Prepare data

To begin with, create a table `test.repository` in your TiDB Serverless cluster and insert some sample data into it. The following example inserts some open source projects developed by PingCAP as data for demonstration purposes.

To execute the SQL statements, you can use the [Chat2Query](/tidb-cloud/explore-data-with-chat2query.md) feature in the [TiDB Cloud console](https://tidbcloud.com/).

```sql
-- Choose the database
USE test;

-- Create the table
CREATE TABLE repository (
        id int NOT NULL PRIMARY KEY AUTO_INCREMENT,
        name varchar(64) NOT NULL,
        url varchar(256) NOT NULL
);

-- Insert some sample data into the table
INSERT INTO repository (name, url)
VALUES ('tidb', 'https://github.com/pingcap/tidb'),
        ('tikv', 'https://github.com/tikv/tikv'),
        ('pd', 'https://github.com/tikv/pd'),
        ('tiflash', 'https://github.com/pingcap/tiflash');
```

> ## 2. Create a Data App

Reviewer 2: ```suggestion
## Step 2. Create a Data App
```

After the data is inserted, navigate to the [**Data Service**](https://tidbcloud.com/console/data-service) page in the [TiDB Cloud console](https://tidbcloud.com/). Create a Data App with the `GET /repositories` endpoint and link the Data App to your TiDB Serverless cluster. The corresponding SQL query for this endpoint is as follows, which fetches all items from the `test.repository` table:

```sql
SELECT * FROM test.repository;
```

For more information, see [Get started with Data Service](/tidb-cloud/data-service-get-started.md)

> ## 3. Generate client code

Reviewer 2: ```suggestion
## Step 3. Generate client code
```

              "200": {
                "content": {
                  "application/json": {
                    "schema": {
                      "$ref": "#/components/schemas/getRepositoriesResponse"
                    }
                  }
                },
                "description": "OK"
              },
    ...
    ```

4. Generate the client code.

    To generate the client code, run the following command:

    ```shell
    yarn run openapi-generator-cli generate -i oas/doc.json --generator-name typescript-fetch -o gen/api
    ```

    This command generates the client code using the `oas/doc.json` specification as input and outputs the client code to the `gen/api` directory.

## 4. Develop your application

You can use the generated client code to develop your application.

1. Create a `.env.local` file in the project directory and set the values to the public key and private key of your Data App.

    ```
    TIDBCLOUD_DATA_SERVICE_PUBLIC_KEY=YOUR_PUBLIC_KEY
    TIDBCLOUD_DATA_SERVICE_PRIVATE_KEY=YOUR_PRIVATE_KEY
    ```

    To create an API key for a Data App, see [Create an API key](/tidb-cloud/data-service-api-key.md#create-an-api-key).

> 2. Replace the content of `app/page.tsx` with the following code, which fetches data from the endpoint and renders it on the page:

Reviewer 2: ```suggestion
2. In the `hello-repos` project directory, replace the content of `app/page.tsx` with the following code, which fetches data from the `GET /repositories` endpoint and renders it on the page:
```

    ```js
    import {DefaultApi, Configuration} from "../gen/api"

    export default async function Home() {
      const config = new Configuration({
        username: process.env.TIDBCLOUD_DATA_SERVICE_PUBLIC_KEY,
        password: process.env.TIDBCLOUD_DATA_SERVICE_PRIVATE_KEY,
      });
      const apiClient = new DefaultApi(config);
      const resp = await apiClient.getRepositories();
      return (
        <main className="flex min-h-screen flex-col items-center justify-between p-24">
          <ul className="font-mono text-2xl">
            {resp.data.rows.map((repo) => (
              <a href={repo.url}>
                <li key={repo.id}>{repo.name}</li>
              </a>
            ))}
          </ul>
        </main>
      )
    }
    ```

    > **Note:**
    >

>     > If your Data App is deployed across multiple regions, you need to configure the endpoint path in `config`. Modify the `config` object as shown below:

Reviewer 2: ```suggestion
    > If your Data App is deployed across multiple regions, you also need to configure the endpoint path in the `config` object as follows:
```

>     >

Reviewer 2: 什么情况下算是 Data App is deployed across multiple regions 呢。

Author: @clark1013 PTAL

    >  ```js
    >  const config = new Configuration({
    >      username: process.env.TIDBCLOUD_DATA_SERVICE_PUBLIC_KEY,
    >      password: process.env.TIDBCLOUD_DATA_SERVICE_PRIVATE_KEY,
    >      basePath: "https://${YOUR_REGION}.data.dev.tidbcloud.com/api/v1beta/app/${YOUR_DATA_APP_ID}/endpoint"
    >    });
    >  ```
    >

>     > Make sure to replace `basePath` with the actual endpoint path of your Data App.

Reviewer 2: ```suggestion
    > Make sure to replace `basePath` with the actual endpoint path of your Data App. To get  `${YOUR_REGION}` and `{YOUR_DATA_APP_ID}`, check the **Endpoint URL** in the endpoint **Properties** pane.
```

---
title: Use the OpenAPI Specification of a Data App with Next.js
summary: Learn how to use the OpenAPI Specification of a Data App to generate client code and develop a Next.js application.
---

# Use the OpenAPI Specification of a Data App with Next.js

This document introduces how to use the OpenAPI Specification of a [Data App](/tidb-cloud/tidb-cloud-glossary.md#data-app) to generate client code and develop a Next.js application.

## Before you begin

Before using OpenAPI Specification with Next.js, make sure that you have the following:

- A TiDB cluster. For more information, see [Create a TiDB Serverless cluster](/tidb-cloud/create-tidb-cluster-serverless.md) or [Create a TiDB Dedicated cluster](/tidb-cloud/create-tidb-cluster.md).

> - [Node.js](https://nodejs.org/en/download).
> - [npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm).
> - [yarn](https://yarnpkg.com/getting-started/install).

Author: ```suggestion
- [Node.js](https://nodejs.org/en/download)
- [npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm)
- [yarn](https://yarnpkg.com/getting-started/install)
```

                  "rows": {
                    "items": {
                      "properties": {
                        "id": {
                          "type": "string"
                        },
                        "name": {
                          "type": "string"
                        },
                        "url": {
                          "type": "string"
                        }
    ...
      "paths": {
        "/repositories": {
          "get": {
            "operationId": "getRepositories",
            "responses": {
              "200": {
                "content": {
                  "application/json": {
                    "schema": {
                      "$ref": "#/components/schemas/getRepositoriesResponse"
                    }
                  }
                },
                "description": "OK"
              },
    ...
    ```

4. Generate the client code:

    ```shell
    yarn run openapi-generator-cli generate -i oas/doc.json --generator-name typescript-fetch -o gen/api
    ```

    This command generates the client code using the `oas/doc.json` specification as input and outputs the client code to the `gen/api` directory.

> ## Step 4. Develop your application

Reviewer 2: ```suggestion
## Step 4. Develop your Next.js application
```

> You can use the generated client code to develop your application.

Reviewer 2: ```suggestion
You can use the generated client code to develop your Next.js application.
```

    TIDBCLOUD_DATA_SERVICE_PRIVATE_KEY=YOUR_PRIVATE_KEY
    ```

    To create an API key for a Data App, see [Create an API key](/tidb-cloud/data-service-api-key.md#create-an-api-key).

2. In the `hello-repos` project directory, replace the content of `app/page.tsx` with the following code, which fetches data from the `GET /repositories` endpoint and renders it:

    ```js
    import {DefaultApi, Configuration} from "../gen/api"

    export default async function Home() {
      const config = new Configuration({
        username: process.env.TIDBCLOUD_DATA_SERVICE_PUBLIC_KEY,
        password: process.env.TIDBCLOUD_DATA_SERVICE_PRIVATE_KEY,
      });
      const apiClient = new DefaultApi(config);
      const resp = await apiClient.getRepositories();
      return (
        <main className="flex min-h-screen flex-col items-center justify-between p-24">
          <ul className="font-mono text-2xl">
            {resp.data.rows.map((repo) => (
              <a href={repo.url}>
                <li key={repo.id}>{repo.name}</li>
              </a>
            ))}
          </ul>
        </main>
      )
    }
    ```

    > **Note:**
    >

>     > If your Data App is deployed across multiple regions, you also need to configure the endpoint path in the `config` object as follows:

Author: ```suggestion
    > If your Data App is deployed across multiple regions, which means it is linked to clusters in different regions, you will see multiple items in the `servers` section of the downloaded OpenAPI Specification file. You also need to configure the endpoint path in the `config` object as follows:
```

Reviewer 2: If your Data App is deployed across multiple regions, which means its linked clusters are hosted in different regions, you will see multiple items in the `servers` section of the downloaded OpenAPI Specification file. In this case, you also need to configure the endpoint path in the `config` object as follows:

    >
    >  ```js
    >  const config = new Configuration({
    >      username: process.env.TIDBCLOUD_DATA_SERVICE_PUBLIC_KEY,
    >      password: process.env.TIDBCLOUD_DATA_SERVICE_PRIVATE_KEY,
    >      basePath: "https://${YOUR_REGION}.data.dev.tidbcloud.com/api/v1beta/app/${YOUR_DATA_APP_ID}/endpoint"
    >    });
    >  ```
    >
    > Make sure to replace `basePath` with the actual endpoint path of your Data App. To get `${YOUR_REGION}` and `{YOUR_DATA_APP_ID}`, check the **Endpoint URL** in the endpoint **Properties** panel.

> ## 5. Preview your application

Reviewer 2: ```suggestion
## 5. Preview your Next.js application
```
